src/libopenrave/kinbodylink.cpp

```diff
@@ -733,54 +733,53 @@ KinBody::Link::GeometryPtr KinBody::Link::GetGeometry(int index)
     return _vGeometries.at(index);
 }
 
-void KinBody::Link::InitGeometries(std::vector<KinBody::GeometryInfoConstPtr>& geometries, bool bForceRecomputeMeshCollision)
+static inline KinBody::Link::Geometry* CreateGeometry(KinBody::LinkPtr plink, const KinBody::GeometryInfoConstPtr& geometryInfoPtr)
 {
-    _vGeometries.resize(geometries.size());
-    for(size_t i = 0; i < geometries.size(); ++i) {
-        _vGeometries[i].reset(new Geometry(shared_from_this(),*geometries[i]));
-        if( bForceRecomputeMeshCollision || _vGeometries[i]->GetCollisionMesh().vertices.size() == 0 ) {
-            if( !bForceRecomputeMeshCollision ) {
-                RAVELOG_VERBOSE("geometry has empty collision mesh\n");
-            }
-            _vGeometries[i]->InitCollisionMesh(); // have to initialize the mesh since some plugins might not understand all geometry types
-        }
-    }
-    _info._mapExtraGeometries.clear();
-    // have to reset the self group! cannot use geometries directly since we require exclusive access to the GeometryInfo objects
-    std::vector<KinBody::GeometryInfoPtr> vgeometryinfos;
-    vgeometryinfos.resize(_vGeometries.size());
-    for(size_t i = 0; i < vgeometryinfos.size(); ++i) {
-        vgeometryinfos[i].reset(new KinBody::GeometryInfo());
-        *vgeometryinfos[i] = _vGeometries[i]->_info;
-    }
-    SetGroupGeometries("self", vgeometryinfos);
-    _Update();
+    return new KinBody::Link::Geometry(plink, *geometryInfoPtr);
 }
 
-void KinBody::Link::InitGeometries(std::list<KinBody::GeometryInfo>& geometries, bool bForceRecomputeMeshCollision)
+static inline KinBody::Link::Geometry* CreateGeometry(KinBody::LinkPtr plink, const KinBody::GeometryInfo& geometryInfo)
 {
+    return new KinBody::Link::Geometry(plink, geometryInfo);
+}
+
+template <typename GeometryIterableT>
+void KinBody::Link::_InitGeometriesInternal(const GeometryIterableT& geometries, bool bForceRecomputeMeshCollision) {
     _vGeometries.resize(geometries.size());
     size_t i = 0;
-    FOREACH(itinfo,geometries) {
-        _vGeometries[i].reset(new Geometry(shared_from_this(),*itinfo));
+    FOREACH(itGeometry, geometries) {
+        _vGeometries[i].reset(CreateGeometry(shared_from_this(), *itGeometry));
```

Comment:
Just curious why a function is necessary here since replacing `new KinBody::Link::Geometry(plink, geometryInfo)` is just as clarifying..

Reply:
The function is necessary to allow the template here to generalize - you can call this with either pointers or non-pointers, so the part of this function that actually changes based on that is extracted (`new KinBody::Link::Geometry(plink, *geometryInfoPtr)` vs `new KinBody::Link::Geometry(plink, geometryInfo)`).

Reply:
okay, i removed the template member function, and now no need for  `CreateGeometry`